README.rst
Introduction
------------

Scribbler is a parallel test runner.

What does it run?
-----------------

Scribbler will pick up any tests where:

* The test method is inside a class
* The test method's name starts with "test\_"

It'll also pick setup methods (methods run before every test) where:

* The setup method is inside a class
* The setup method's name is either "setUp", "setup" or "set_up"

It'll pick teardown methods as well (methods run after every test) where:

* The teardown method is inside a class
* The teardown method's name is either "tearDown", "teardown" or "tear_down"

Will I get any results
----------------------

It's hard for me to spoil the surprise of running, but if you run::

    scribbler_console

Scribbler will try to run all the tests in the current directory (and sub-directories) with the default settings.

For more info on the options run the help with::

    scribbler_console -h

Why Scribbler?
--------------

Ernest Scribbler is the inventor of the Funniest Joke in the World (Monthy Python's sketch).

We love the sketch, hence the name.


debian/changelog
python-scribbler (0.2) alpha; urgency=low

  * Initial release.

 -- Gabriel Falcão Gonçalves de Moura <gabriel.falcao@corp.globo.com>  Thu, 07 May 2009 11:00:51 -0300


tests/test_test_parser.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import sys
from os.path import dirname, abspath, join
test_path = abspath(dirname(__file__))
root_path = abspath(join(dirname(__file__), "../"))
sys.path.insert(0, root_path)

from scribbler import TestParser
from samples.sample_test import SampleTest

def test_test_parser_retrieves_something():
    parser = TestParser(tests_dir = join(test_path, "samples"))

    parser.parse()
    tests = parser.tests
    assert tests is not None

def test_test_parser_retrieves_one_test():
    parser = TestParser(tests_dir = join(test_path, "samples"))

    parser.parse()
    tests = parser.tests
    assert len(tests) == 1, "The length should be 1 but was %d" % len(tests)

def test_test_parser_retrieves_sample_test():
    parser = TestParser(tests_dir = join(test_path, "samples"))

    parser.parse()
    tests = parser.tests
    
    assert tests[0].tests[0].__name__ == "test_pass", \
                "The test should be test_pass but it is a %s instead." % \
                tests[0].tests[0].__name__



tests/scribbler/__init__.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import sys
from os.path import dirname, abspath, join
root_path = abspath(join(dirname(__file__), "../../"))
sys.path.insert(0, root_path)

from Queue import Queue
from threading import Thread

class TestRunner(object):
    Success = "SUCCESS"
    Failure = "FAILURE"

    def __init__(self, test_suite, working_threads = 5):
        self.test_suite = test_suite
        self.working_threads = int(working_threads)
        self.test_queue = Queue()
        self.results = TestResult()
        for item in self.test_suite:
            self.test_queue.put(item)

    def worker(self):
        while True:
            item = self.test_queue.get()
            try:
                result = item()
                self.results.append(self.Success, None)
            except Exception, err:
                self.results.append(self.Failure, unicode(err))
            self.test_queue.task_done()

    def start_processes(self):
        for i in range(self.working_threads):
            t = Thread(target=self.worker)
            t.setDaemon(True)
            t.start()

    def run(self):
        self.start_processes()

        self.test_queue.join()

        return self.results

class TestResult(object):
    def __init__(self):
        self.results = []
        self.errors = []

    def append(self, test_result, error):
        self.results.append(test_result)
        self.errors.append(error)


tests/console_tests/tests.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import unittest
import time

class TestConsole(unittest.TestCase):
    def test_wait_for_3_seconds(self):
        time.sleep(3)
        assert True
    def test_wait_for_2_seconds(self):
        time.sleep(2)
        assert True
    def test_wait_for_1_seconds(self):
        time.sleep(1)
        assert True
    def test_wait_for_FourthOfA_seconds(self):
        time.sleep(0.25)
        assert True
    def test_wait_for_Half_seconds(self):
        time.sleep(0.5)
        assert True
    def test_wait_for_ThirdOfA_seconds(self):
        time.sleep(0.33)
        assert True


scribbler/scribbler_console.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

u"""scribbler - BDD Acceptance testing

Example usage
=============

    python scribbler.py

:author: `Bernardo Heynemann <mailto:heynemann@gmail.com>`__
"""

__revision__ = "$Id$"
__docformat__ = 'restructuredtext en'

import os
import sys
import optparse
from os.path import join, dirname, abspath
from scribbler import TestParser, TestRunner, TestResult
import time

def main():
    """ Main function - parses args and runs action """
    parser = optparse.OptionParser(usage="%prog or type %prog -h (--help) for help", description=__doc__, version="%prog" + __revision__)
    parser.add_option("-d", "--dir", dest="dir", default=None, help="Directory to run the tests in. If none specified it runs in the current directory.")
    parser.add_option("-p", "--pattern", dest="pattern", default="*.py", help="Pattern of files to use for retrieving tests.")
    parser.add_option("-r", "--recursive", dest="recursive", default="true", help="Specifies whether sub-folders of the 'dir' parameter should be parsed as well. Defaults to 'true' (set to 'false' to parse only the tests in the specified dir)")
    parser.add_option("-t", "--threads", dest="threads", default="5", help="Number of threads to run tests with [default: %default].")

    (options, args) = parser.parse_args()

    tests_dir = options.dir and abspath(options.dir) or abspath(os.curdir)

    start_time = time.time()

    parser = TestParser(tests_dir=tests_dir, pattern = options.pattern)
    parser.parse()
    
    test_suite = parser.tests
    
    if not test_suite:
        print "No tests found under %s with the %s pattern" % (tests_dir, options.pattern)
        sys.exit(0)
    
    runner = TestRunner(test_suite=test_suite, pattern=options.pattern, working_threads=int(options.threads))
    
    def print_success(name, test):
        sys.stdout.write(".")
        sys.stdout.flush()
    def print_failure(name, test, message):
        sys.stdout.write("F")
        sys.stdout.flush()

    runner.test_successful = print_success
    runner.test_failed = print_failure
        
    results = runner.run()

    print

    print_tests(results)

    print "-" * 80
    print "Test run successful"
    show_time(start_time)
    print "-" * 80

    sys.exit(0)

def print_tests(results):
    print "Tests:"
    print "-" * 80    
    i = 1
    for name, result, error in results.results:
        if error:
            print "%d - %s - %s\n%s\n\n" % (i, result.lower(), name, error)
        else:
            print "%d - %s - %s" % (i, result.lower(), name)        
        i+=1
    print 

def show_time(start_time):
    period = (time.time() - start_time)
    print "%0.2f s" % period
    
if __name__ == "__main__":
    main()



scribbler/__init__.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import sys
from os.path import dirname, abspath, join, split, splitext
root_path = abspath(join(dirname(__file__), "../../"))
sys.path.insert(0, root_path)
from Queue import Queue
from threading import Thread
import unittest
import re
import inspect
import time
import traceback

import locator

class TestParser(object):
    def __init__(self, tests_dir, pattern="*.py"):
        self.tests_dir = tests_dir
        self.pattern = pattern
        self.tests = {}

    def parse(self):
        tests = []
        for filename in locator.locate(self.pattern, self.tests_dir):
            filedir = dirname(filename)
            filename = splitext(split(filename)[1])[0]
            if re.match("^[a-zA-Z]\w+$", filename):
                sys.path.append(filedir)
                try:
                    mod = __import__('%s' % filename)
                except ImportError, err:
                    raise SyntaxError("Error importing '%s'. Error: %s" % (filename, str(err)))
                sys.path.pop()
                tests.extend(self.load_fixture_from_module(mod))
        self.tests = tests

    def load_fixture_from_module(self, mod):
        fixtures = []
        for key, obj in inspect.getmembers(mod):
            if not inspect.isclass(obj): continue
            fixture = TestFixture()
            fixture.test_case = obj
            for key in dir(obj):
                method = getattr(obj, key)
                if not (inspect.ismethod(method) or inspect.isfunction(method)):
                    continue
                if key.startswith("test_"):
                    fixture.tests.append(method)
                    continue
                if key in ["setUp", "setup", "set_up"]:
                    if fixture.setup is None:
                        fixture.setup = method
                    continue
                if key in ["tearDown", "teardown", "tear_down"]:
                    if fixture.teardown is None:
                        fixture.teardown = method
                    continue
            if fixture.tests:
                fixtures.append(fixture)
        
        return fixtures

class TestRunner(object):
    before_test = None
    test_successful = None
    test_failed = None
    tests_executing = 0

    def __init__(self, test_suite, pattern="*.py", working_threads = 5):
        self.test_suite = test_suite
        self.working_threads = int(working_threads)
        self.test_queue = Queue()
        self.results = TestResult()
        for test_fixture in [suite for suite in self.test_suite if suite is not None]:
            setup = test_fixture.setup
            teardown = test_fixture.teardown
            test_case = test_fixture.test_case
            for test in test_fixture.tests: 
                self.test_queue.put((test_case, setup, teardown, test))

    def worker(self):
        while True:
            test_case_type, setup_method, teardown_method, test_method = self.test_queue.get()
            self.tests_executing += 1
            name = test_method.__name__
            try:
                if self.before_test:
                    self.before_test(name, test_method)
                if unittest.TestCase in test_case_type.__bases__:
                    test_case = test_case_type(name)
                else:
                    test_case = test_case_type()
                if setup_method: setup_method(test_case)
                try:
                    result = test_method(test_case)
                finally:
                    if teardown_method: teardown_method(test_case)
                if self.test_successful:
                    self.test_successful(name, test_method)
                self.results.append(name, TestResult.Success, None)
            except Exception, err:
                if self.test_failed:
                    self.test_failed(name, test_method, err)
                self.results.append(name, TestResult.Failure, traceback.format_exc())
            self.tests_executing -= 1
            self.test_queue.task_done()

    def start_processes(self):
        for i in range(self.working_threads):
            t = Thread(target=self.worker)
            t.setDaemon(True)
            t.start()

    def run(self):
        self.start_processes()

        try:
            time.sleep(0.5)
            while (self.tests_executing > 0):
                time.sleep(0.2)
        except KeyboardInterrupt:
            print "Test Run interrupted by the user..."

        return self.results

class TestResult(object):
    Success = "SUCCESS"
    Failure = "FAILURE"

    def __init__(self):
        self.results = []

    def append(self, name, test_result, error):
        self.results.append((name, test_result, error))
    
    def get_status(self):
        status = self.Success
        for name, test_result, error in self.results:
            if error:
                status = self.Failure
        return status

class TestFixture(object):
    def __init__(self):
        self.test_case = None
        self.setup = None
        self.teardown = None
        self.tests = []
    
